src/porepy/numerics/ad/functions.py

```diff
@@ -401,10 +399,22 @@ def maximum(var_0: pp.ad.AdArray, var_1: pp.ad.AdArray | np.ndarray) -> pp.ad.Ad
     else:
         max_jac[inds] = jacs[1][inds]
 
-    return pp.ad.AdArray(max_val, max_jac)
+    return AdArray(max_val, max_jac)
+
 
+@overload
+def characteristic_function(tol: float, var: np.ndarray | float) -> np.ndarray:
+    ...
 
-def characteristic_function(tol: float, var: pp.ad.AdArray):
+
+@overload
+def characteristic_function(tol: float, var: AdArray) -> AdArray:
+    ...
+
+
+def characteristic_function(
+    tol: float, var: Union[AdArray, np.ndarray, float]
+) -> np.ndarray | AdArray:
```

Comment:
This is the place where you highlighted the choice between overload and TypeVar? If so, I much prefer TypeVar if that is fine with you.

Reply:
Meaning that we change the return type to float, rather than bool? If yes, do you want me to do it?

Reply:
Yes and yes, please.